### ISSUE TITLE
[ADAP-346] job_execution_timeout_seconds config variable not used when running python models on dataproc. Python models timeout after 900 seconds

### ISSUE THREAD

**Issue by toblo956** [dbt-labs/dbt-bigquery#577]
### Is this a new bug in dbt-bigquery?

- [X] I believe this is a new bug in dbt-bigquery
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

We are currently running 3 python models in our DAG. Both locally and in Airflow (once a month). One of our python models use a lot of data and takes some time to complete (normally 20-30 min). When we were using dbt-core 1.3.0 in production there were no issues until the dataproc bug was introduced [bug](https://github.com/dbt-labs/dbt-bigquery/issues/531). This forced us to upgrade dbt-bigquery and dbt-core. 

DBT Python models running in dataproc time out after 15 minutes even though we have set 
`job_execution_timeout_seconds: 7200` - both in the repo and in our ~/.dbt folder. We have also tried adding it into the model-specific .yml file but without success.

Running dbt --debug gave us where the 900 default timeout is coming from. The google package python-core-api sets the default timeout:

python3.9/site-packages/google/api_core/future/polling.py:
`# Default polling configuration
DEFAULT_POLLING = retries.Retry(
    predicate=POLLING_PREDICATE,
    initial=1.0,  # seconds
    maximum=20.0,  # seconds
    multiplier=1.5,
    timeout=900,  # seconds
)`

My question is: why isn't our timeout being used for python models? And what is the best way to fix this? Right now we are editing the polling.py file to timeout after 7200 seconds which works but isn't feasible long-term.

### Expected Behavior

dbt python models running in dataproc use the `job_execution_timeout_seconds: 7200` variable set in our profiles.yml

### Steps To Reproduce

1. dbt-bigquery 1.3.1 dbt-core 1.3.1 (dbt-bigquery 1.4.x has bugs related to incremental models being materialized as tables)
2. Set `job_execution_timeout_seconds: 7200`
3. Run a python model which takes longer than 15 min to complete

### Environment

```markdown
- OS: Mac OS Ventura 13.0
- Python: 3.9.6
- dbt-core: 1.3.1
- dbt-bigquery: 1.3.1
```

**Comment by Fleid** [comment]
Hi @toblo956,

Can you confirm you get a incorrect value when you execute `dbt debug`?

I tried reproducing and couldn't:

```bash
22:20:21  Running with dbt=1.3.1
dbt version: 1.3.1
python version: 3.10.5
python path: /Users/fleide/Documents/Repos/adap-346/.venv/bin/python3
os info: macOS-13.2.1-arm64-arm-64bit
Using profiles.yml file at /Users/fleide/.dbt/profiles.yml
Using dbt_project.yml file at /Users/fleide/Documents/Repos/adap-346/adap_346/dbt_project.yml

Configuration:
  profiles.yml file [OK found and valid]
  dbt_project.yml file [OK found and valid]

Required dependencies:
 - git [OK found]

Connection:
  method: oauth
  database: ...
  schema: ...
  location: US
  priority: interactive
  timeout_seconds: 7200
  maximum_bytes_billed: None
  execution_project: ...
  job_retry_deadline_seconds: None
  job_retries: 1
  job_creation_timeout_seconds: None
  job_execution_timeout_seconds: 7200
  gcs_bucket: None
  Connection test: [OK connection ok]
```

My profile is straightforward:
```YAML
adap_346:
  outputs:
    dev:
      dataset: ...
      job_execution_timeout_seconds: 7200
      job_retries: 1
      location: US
      method: oauth
      priority: interactive
      project: ...
      threads: 1
      type: bigquery
  target: dev
```

What connection method are you using?

**Comment by Fleid** [comment]
By the way, do you have a list of the bugs that are preventing you from upgrading to 1.4?

**Comment by toblo956** [comment]
> Hi @toblo956,
> 
> Can you confirm you get a incorrect value when you execute `dbt debug`?
> 
> I tried reproducing and couldn't:
> 
> ```shell
> 22:20:21  Running with dbt=1.3.1
> dbt version: 1.3.1
> python version: 3.10.5
> python path: /Users/fleide/Documents/Repos/adap-346/.venv/bin/python3
> os info: macOS-13.2.1-arm64-arm-64bit
> Using profiles.yml file at /Users/fleide/.dbt/profiles.yml
> Using dbt_project.yml file at /Users/fleide/Documents/Repos/adap-346/adap_346/dbt_project.yml
> 
> Configuration:
>   profiles.yml file [OK found and valid]
>   dbt_project.yml file [OK found and valid]
> 
> Required dependencies:
>  - git [OK found]
> 
> Connection:
>   method: oauth
>   database: ...
>   schema: ...
>   location: US
>   priority: interactive
>   timeout_seconds: 7200
>   maximum_bytes_billed: None
>   execution_project: ...
>   job_retry_deadline_seconds: None
>   job_retries: 1
>   job_creation_timeout_seconds: None
>   job_execution_timeout_seconds: 7200
>   gcs_bucket: None
>   Connection test: [OK connection ok]
> ```
> 
> My profile is straightforward:
> 
> ```yaml
> adap_346:
>   outputs:
>     dev:
>       dataset: ...
>       job_execution_timeout_seconds: 7200
>       job_retries: 1
>       location: US
>       method: oauth
>       priority: interactive
>       project: ...
>       threads: 1
>       type: bigquery
>   target: dev
> ```
>
> What connection method are you using?

This is what I get when I run dbt debug:

```
Running with dbt=1.3.1
dbt version: 1.3.1
python version: 3.9.6
os info: macOS-13.0-arm64-arm-64bit
Using profiles.yml file at ~/.dbt/profiles.yml
Using dbt_project.yml file at dbt-us/dbt_project.yml

Configuration:
  profiles.yml file [OK found and valid]
  dbt_project.yml file [OK found and valid]

Required dependencies:
 - git [OK found]

Connection:
  method: oauth
  database: temp
  schema: ...
  location: EU
  priority: interactive
  timeout_seconds: 7200
  execution_project: ...
  job_retry_deadline_seconds: None
  job_retries: 2
  job_creation_timeout_seconds: None
  job_execution_timeout_seconds: 7200
  gcs_bucket: ...
  Connection test: [OK connection ok]

All checks passed!
```

This is the profile I am using:

```
outputs:
    main-project-stage:
      type: bigquery
      method: oauth
      project: ...
      dataset: temp
      job_retries: 2
      location: EU
      threads: 1
      timeout_seconds: 7200
      priority: interactive
      gcs_bucket: ...
      dataproc_region: europe-west1
```

So by just looking at the config and running debug, it looks like it is working. However, the timeout that is actually being used in dataproc is set by google-api-core (it is default 900). One way to reproduce this error is to change this variable to something low, such as 15 seconds and see that it is used despite providing 7200 seconds from the config.

I edited in my locally installed google-api-core: path_to_venv/lib/python3.9/site-packages/google/api_core/future/polling.py

# Default polling configuration
DEFAULT_POLLING = retries.Retry(
    predicate=POLLING_PREDICATE,
    initial=1.0,  # seconds
    maximum=20.0,  # seconds
    multiplier=1.5,
    timeout=7200 # 900 -> 7200. Try 15 seconds to verify this being used instead of job_execution_timeout from config
)

**Comment by toblo956** [comment]
> By the way, do you have a list of the bugs that are preventing you from upgrading to 1.4?

We have only noticed one bug, but it is quite significant. When we run our python models with dbt-bigquery 1.4 all models that are configured as incremental will be materialized as tables. I lifted this issue in the dbt-slack and there were others who had encountered the same problem.

**Comment by mzemp** [comment]
> By the way, do you have a list of the bugs that are preventing you from upgrading to 1.4?

You can find a description of an issue with an indentation error in `dbt-bigquery 1.4.1` [here](https://github.com/dbt-labs/dbt-bigquery/issues/581#issuecomment-1456214428).

**Comment by Fleid** [comment]
I tried:
- Polling timeout 1 sec, dbt timeout 7200 sec : successfully ran a 17 sec model
- Polling timeout 7200 sec, dbt timeout 1 sec : failed to run at 1 sec

I'm running Python 3.10.5 and google-api-core 2.11.0

Which version of the google api are you using?

**Comment by toblo956** [comment]
I was using python 3.9.6. I also tried with python 3.10.0 and my colleague tried with 3.10.5. We are both using google-api-core 2.11.0. One thing to add is that we are running dataproc serverless. Maybe that is why it's different from your results?

- Polling timeout 10 sec, dbt timeout 7200 sec, failed to run after 10 sec
- Polling timeout 7200, dbt timeout 1 sec, successfully ran after 24 min

**Comment by Fleid** [comment]
I had messed up my setup, and was testing a project with a SQL model instead...
I can confirm the behavior is only observed on Python models.

With 
- Polling timeout 2 sec, dbt timeout 7200 sec, **in the same run**
  - 1 of 2 OK created sql table model dbt_feiden.my_first_dbt_model ................ [CREATE TABLE (313.8m rows, 35.7 GB processed) in 15.82s]
  - 2 of 2 ERROR creating python table model dbt_feiden.my_python_model ............ [ERROR in 3.56s] : **Operation did not complete within the designated timeout of 2 seconds.**

**Comment by glevineLeap** [comment]
Does anyone have a workaround for this until the bug fix is released? We're running 1.4 (latest) on dbt Cloud but will soon be upgrading to 1.5.

**Comment by nickozilla** [comment]
Is this still being worked on? My team would also like to see this progressed if possible

**Comment by Fleid** [comment]
We want to have that included in 1.6.0 but we may not be able to.
It's near the top of the pile, but the pile is long :/

**Comment by dataders** [comment]
linking #263 for posterity
